tools/build_defs/repo/git.bzl

```diff
@@ -49,29 +107,45 @@ def _clone_or_update(ctx):
                   ctx.attr.strip_prefix if ctx.attr.strip_prefix else "None",
               ))
     bash_exe = ctx.os.environ["BAZEL_SH"] if "BAZEL_SH" in ctx.os.environ else "bash"
-    st = ctx.execute([bash_exe, "-c", """
-cd {working_dir}
-set -ex
-( cd {working_dir} &&
-    if ! ( cd '{dir_link}' && [[ "$(git rev-parse --git-dir)" == '.git' ]] ) >/dev/null 2>&1; then
-      rm -rf '{directory}' '{dir_link}'
-      git clone '{shallow}' '{remote}' '{directory}' || git clone '{remote}' '{directory}'
-    fi
-    git -C '{directory}' reset --hard {ref} || \
-    ((git -C '{directory}' fetch '{shallow}' origin {ref}:{ref} || \
-      git -C '{directory}' fetch origin {ref}:{ref}) && git -C '{directory}' reset --hard {ref})
-      git -C '{directory}' clean -xdf )
-  """.format(
-        working_dir = ctx.path(".").dirname,
-        dir_link = ctx.path("."),
-        directory = directory,
-        remote = ctx.attr.remote,
-        ref = ref,
-        shallow = shallow,
-    )], environment = ctx.os.environ)
+    git_cache = ctx.os.environ.get("BAZEL_GIT_REPOSITORY_CACHE")
+
+    if git_cache:
+        _setup_cache(ctx, git_cache)
+        st = _get_repository_from_cache(ctx, directory, ref, git_cache)
+
+        if st.return_code:
+            _populate_cache(ctx, git_cache, remote_name, ref, shallow)
```

Comment:
This approach has the problem, that `git fetch` is called unconditionally, and `git fetch` tries to contact the remote site, even if we only need a commit already in cache. That means, we cannot use the cache when being offline, which would be a main use case for a cache.

This can be avoided by first trying to check out the correct commit and only if that fails, trying to update the cache.